### PR TITLE
8314236: Overflow in Collections.rotate

### PR DESCRIPTION
**Pull request by nikita-sakharin**
`Collections.rotate` method contains a bug. This method throws IndexOutOfBoundsException on arrays larger than $2^{30}$ elements. The way to reproduce:
```
final int size = (1 << 30) + 1;
final List<Byte> list = new ArrayList<>(size);
for (int i = 0; i < size; ++i)
    list.add((byte) 0);
Collections.rotate(list, size - 1);
````
Output:
```Exception in thread "main" java.lang.IndexOutOfBoundsException: Index -2147483648 out of bounds for length 1073741825```

In that case private method `Collections.rotate1` will be called. And the line:
`i += distance;`
will cause overflow. I fixed this method and wrote a test for it.

I've signed the Oracle Contributor Agreement, but I don't have permission to raise a bug in the JDK Bug System.

Kindly ask you to raise a bug.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8314236](https://bugs.openjdk.org/browse/JDK-8314236): Overflow in Collections.rotate (**Bug** - P4)

### Reviewers
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**)
 * [Stuart Marks](https://openjdk.org/census#smarks) (@stuart-marks - **Reviewer**)

### Contributors
 * Nikita Sakharin `<17588081+nikita-sakharin@users.noreply.github.com>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15270/head:pull/15270` \
`$ git checkout pull/15270`

Update a local copy of the PR: \
`$ git checkout pull/15270` \
`$ git pull https://git.openjdk.org/jdk.git pull/15270/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15270`

View PR using the GUI difftool: \
`$ git pr show -t 15270`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15270.diff">https://git.openjdk.org/jdk/pull/15270.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15270#issuecomment-1678063621)

### CODE REVIEW

**Comment by shipilev** [comment]
Submitted: [JDK-8314236](https://bugs.openjdk.org/browse/JDK-8314236)

Please change the PR synopsis to: "8314236: Overflow in Collections.rotate".

Also go to https://github.com/nikita-sakharin/jdk/actions, and enable testing workflows.

**Review by shipilev — COMMENTED** [review]
Superficially, this looks okay. Some OpenJDK code/test style trivia:

File: test/jdk/java/util/Collections/RotateHuge.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8314236
+ * @summary Overflow in Collections.rotate
```
[the remark this answers is not in the document]
Comment (stuart-marks) [reply test/jdk/java/util/Collections/RotateHuge.java:27]: @nikita-sakharin

Thanks for finding this bug and offering to fix it! (And @shipilev thanks for your assistance on this.)

Putting the test into a separate JVM will work, but I don't think it's necessary to actually allocate the space. The test is only testing the indexes sent to `get` and `set` on the list, and it doesn't actually verify the contents of the list. (Presumably that's done by other tests.) Therefore it should be possible to create a "virtual" list of a given size that checks that the indexes are all in bounds but that doesn't actually store any elements. It should be fairly straightforward to do this by subclassing AbstractList and overriding a few methods.

The advantage of not actually allocating 4G of memory is that it makes it easier to run a bunch of cases that test the boundary conditions. In fact I'd like to see that in the test, as opposed to testing this one case.

**Comment by stuart-marks** [comment]
@nikita-sakharin

Thanks for the updates. With the "Mock List" implementation we can run the test in-JVM and we can avoid allocating several GB of memory. Great!

The implementation logic in the `rotate1` method looks correct.

Now that an individual test case is much less expensive, it becomes feasible to add multiple test cases. In particular, for this kind of testing of arithmetic errors, I like to test a variety of edge cases. The one test case you have is for size=`(1 << 30) - 1` and distance=`(1 << 30)`. It would be good to have a few other cases where the existing code fails and where the modified code should pass. I was able to come up with a few examples quickly:

    size                 distance
    Integer.MAX_VALUE    2
    Integer.MAX_VALUE    Integer.MIN_VALUE
    Integer.MAX_VALUE    Integer.MAX_VALUE - 1

Please add these cases, and any others that you think might be interesting.

**Comment by nikita-sakharin** [comment]
@stuart-marks
Thank you for your review.

Test was failing due to [JDK-8313701](https://bugs.openjdk.org/browse/JDK-8313701). So, I pulled changes from the `master` and squash my commits.

I increased test coverage with respect to your advice. All tests passed successfully.

Awaiting the next step of your review.

**Comment by stuart-marks** [comment]
(The bot has already admonished you about avoiding rebasing. In practice merging works well and is helpful so that reviewers can see incremental changes.)

The additional test cases look good. Sorry I forgot to mention this previously: it would be good to have a comment in the test that says that the test is _only_ testing the index computations and not the actual rotation of elements.

Meanwhile, I'll pull in the changes from your branch and run them through our internal build system, which is more reliable (and comprehensive) than GHA.

**Comment by nikita-sakharin** [comment]
@stuart-marks

Thank you for your review!

I added comment for the test.

**Review by stuart-marks — APPROVED** [review]
Thanks for the update. Test run looks good! Approved.

**Comment by nikita-sakharin** [comment]
@shipilev

Stuart Marks has approved my changes. You are reviewer for that PR too.

Could I kindly ask you to review the PR, please?

**Comment by nikita-sakharin** [comment]
/contributor add Nikita Sakharin <17588081+nikita-sakharin@users.noreply.github.com>

**Comment by nikita-sakharin** [comment]
Aleksey Shipilëv (@shipilev), kindly remind you about the PR.

Could I ask you to review it, please?

**Review by shipilev — APPROVED** [review]
OK, looks reasonable.

**Comment by nikita-sakharin** [comment]
Aleksey Shipilëv (@shipilev), Stuart Marks (@stuart-marks), thank you for your reviews. Since both reviewers approved the PR, it is ready to be integrated.

I have already executed `/integrate` command. Now I am awaiting one of you to execute `/sponsor`  command. So I kindly ask you to do that.

**Comment by shipilev** [comment]
I would like to push this early next week.

**Pull request closed without merging**